edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesServiceBuilder.cs

```diff
@@ -0,0 +1,163 @@
+    using DockerModels = global::Docker.DotNet.Models;
+
+    public class KubernetesServiceBuilder<TConfig>
+    {
+        readonly string defaultMapServiceType;
+
+        public KubernetesServiceBuilder(string defaultMapServiceType)
+        {
+            this.defaultMapServiceType = defaultMapServiceType;
+        }
+
+        public Option<V1Service> GetServiceFromModule(Dictionary<string, string> labels, KubernetesModule<TConfig> module, IModuleIdentity moduleIdentity)
+        {
+            var portList = new List<V1ServicePort>();
+            Option<Dictionary<string, string>> serviceAnnotations = Option.None<Dictionary<string, string>>();
+            bool onlyExposedPorts = true;
+            if (module is IModule<AgentDocker.CombinedDockerConfig> moduleWithDockerConfig)
+            {
+                if (moduleWithDockerConfig.Config.CreateOptions?.Labels != null)
+                {
+                    // Add annotations from Docker labels. This provides the customer a way to assign annotations to services if they want
+                    // to tie backend services to load balancers via an Ingress Controller.
+                    var annotations = new Dictionary<string, string>();
+                    foreach (KeyValuePair<string, string> label in moduleWithDockerConfig.Config.CreateOptions?.Labels)
+                    {
+                        annotations.Add(KubeUtils.SanitizeAnnotationKey(label.Key), label.Value);
+                    }
+
+                    serviceAnnotations = Option.Some(annotations);
+                }
+
+                // Handle ExposedPorts entries
+                if (moduleWithDockerConfig.Config?.CreateOptions?.ExposedPorts != null)
+                {
+                    // Entries in the Exposed Port list just tell Docker that this container wants to listen on that port.
+                    // We interpret this as a "ClusterIP" service type listening on that exposed port, backed by this module.
+                    // Users of this Module's exposed port should be able to find the service by connecting to "<module name>:<port>"
+                    this.GetExposedPorts(moduleWithDockerConfig.Config.CreateOptions.ExposedPorts)
+                        .ForEach(
+                            exposedList =>
+                                exposedList.ForEach((item) => portList.Add(new V1ServicePort(item.Port, name: $"ExposedPort-{item.Port}-{item.Protocol.ToLower()}", protocol: item.Protocol))));
+                }
+
+                // Handle HostConfig PortBindings entries
+                if (moduleWithDockerConfig.Config?.CreateOptions?.HostConfig?.PortBindings != null)
+                {
+                    foreach (KeyValuePair<string, IList<DockerModels.PortBinding>> portBinding in moduleWithDockerConfig.Config?.CreateOptions?.HostConfig?.PortBindings)
+                    {
+                        string[] portProtocol = portBinding.Key.Split('/');
```

Comment:
We are using parse port and protocol in several places. Should we create a new class that represents this pair and method `Parse` to encapsulate parsing logic inside it? It should also help with testing

Reply:
I thought that I did?